staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go

```diff
@@ -270,21 +280,179 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
 						return errs, -1
 					} else {
 						klog.V(2).ErrorS(msgErr, "messageExpression evaluation failed")
-						errs = append(errs, field.Invalid(fldPath, sts.Type, ruleMessageOrDefault(rule)))
+						errs = append(errs, fieldErrorForReason(fldPath, sts.Type, ruleMessageOrDefault(rule), rule.Reason))
 						remainingBudget = newRemainingBudget
 					}
 				} else {
-					errs = append(errs, field.Invalid(fldPath, sts.Type, messageExpression))
+					errs = append(errs, fieldErrorForReason(fldPath, sts.Type, messageExpression, rule.Reason))
 					remainingBudget = newRemainingBudget
 				}
 			} else {
-				errs = append(errs, field.Invalid(fldPath, sts.Type, ruleMessageOrDefault(rule)))
+				errs = append(errs, fieldErrorForReason(fldPath, sts.Type, ruleMessageOrDefault(rule), rule.Reason))
 			}
 		}
 	}
 	return errs, remainingBudget
 }
 
+// validFieldPath returns a valid field path with whitespace trimmed if the given field path is valid.
+func validFieldPath(fieldPath string, pathOfFieldPath *field.Path, schema *schema.Structural) (validFieldPath *field.Path, allErrs field.ErrorList) {
```

Comment:
This looks like basically the same parser as ` validateFieldPath()`. Can we consolidate the code somehow so that we don't have the parser duplicated and risk drift?

Reply:
Updated